Make error messages specific

kimlisa opened the pull request:
resolves https://github.com/gravitational/teleport/issues/6046

#### Description
Note: there is a bug with trace marshalling proxyErr: https://github.com/gravitational/teleport/pull/6108#discussion_r599795110

Purpose is to allow users with admin privilege that are able to view audit logs, to be able to debug SSO login failures from the UI as much as possible:
- emit audit events when there are failures from creating sso auth requests
- since we no longer expose any specific errors when users fail to login from UI, returned error messages are more specific so that it is clear what the error is when viewing audit logs
- return generic `login failed` message for sso CLI login failures (which are errors from creating an auth request) similar in the UI dashbaord where we also display generic error message

kimlisa committed:
Make error messages specific

a-palchikov commented on lib/auth/auth_test.go:
```diff
@@ -1085,6 +1085,28 @@ func TestU2FSignChallengeCompat(t *testing.T) {
+	mockE := &events.MockEmitter{}
+
+	emitSSOLoginFailureEvent(context.Background(), mockE, "test", trace.BadParameter("some error"))
+	event := mockE.LastEvent().(*events.UserLogin)
```
Instead of type assertion (which might potentially trigger a panic), assert the equivalence with `require`, e.g. it should suffice to:
```go
require.Equal(t, mockE.LastEvent(), &events.UserLogin{
	Metadata: events.Metadata{
		Type: events.UserLoginEvent,
		Code: events.UserSSOLoginFailureCode,
	},
	Method: "test",
	Status: events.Status{
		Success:     false,
		Error:       "some error",
		UserMessage: "some error",
	},
})
```
and matching on `Method` will not be necessary since it's also part of the expectation.

kimlisa replied:
revised

a-palchikov commented on lib/auth/oidc.go:
```diff
@@ -810,21 +809,21 @@ func (a *Server) getClaims(oidcClient *oidc.Client, connector services.OIDCConne
 	if idsub != uisub {
-		log.Debugf("OIDC claim subjects don't match '%v' != '%v'.", idsub, uisub)
-		return nil, trace.BadParameter("invalid subject in UserInfo")
+		log.Debugf("OIDC claim subjects does not match '%v' != '%v'.", idsub, uisub)
```
I think it was grammatically correct before?

kimlisa replied:
reverted

a-palchikov commented on lib/utils/cli.go:
```diff
@@ -196,8 +197,15 @@ func UserMessageFromError(err error) string {
 		} else {
-			fmt.Fprintln(&buf, AllowNewlines(err.Error()))
+			strErr := err.Error()
+			// Error can be of type trace.proxyError where error message didn't get captured.
```
This is somewhat misleading. Did you have a specific error that did not transport the message over the wire? All errors coming from the server over HTTP will be of type [trace.proxyError](https://github.com/gravitational/trace/blob/bbab4b868a4a8a002ffc76d005d53996506ef059/trace.go#L594) and it's supposed to transport the message with it - if that's not the case here, it's probably a bug to be fixed.

kimlisa replied:
I was trying to purposely cause errors when logging in with sso through CLI and I found errors came back as empty strings. I traced into it and I thought maybe it was by design to prevent sensitive error messages from auth from being exposed?

This is how errors get marshalled: https://github.com/gravitational/trace/blob/master/httplib.go#L87
And this is how errors get unmarshalled: https://github.com/gravitational/trace/blob/master/httplib.go#L59
And this is how requests are made in both client and proxy: https://github.com/gravitational/teleport/blob/master/lib/auth/clt.go#L251

Given tsh client -> proxy -> auth make request
First marshal/unmarshal auth to proxy: when error from auth gets marshalled,  the error is of type `TraceErr`, and when it gets unmarshalled, error is unmarshaled into `TraceErr` and wrapped in `proxyErr`

Second marshal/unmarshal proxy to client: When this proxyErr gets marshalled, b/c it is not a type of `TraceErr`, everything gets stripped except error field, and when this gets unmarshalled it gets unmarshalled into trace error struct that only has `message` as its field, so the error field never gets saved and message is empty.

a-palchikov replied:
Ok, I see. This a bug in trace and should be definitely fixed so it's possible to send errors over the wire multiple times. I'll see how complicated this would be.

kimlisa replied:
Since this is a bug, I went back to all console logins `github`, `oidc`, `saml`, in the webapi layer and made sure to return a generic error message for all errors to hide sensitive data. This is what we did for UI dashboard logins (and validating auth callbacks) where any errors encountered we redirected users to a generic error page with direction to check teleports log: https://github.com/gravitational/teleport/pull/6108/commits/50078d3cbbaf743e457a719b93b4ce81b5c203e7

kimlisa committed:
Remove double emit and return generic error message
- Remove inner emit event for createOIDCClient to prevent double emits
  on error.
- Return generic error message for console login errors

kimlisa commented on lib/auth/oidc.go:
```diff
@@ -103,25 +103,7 @@ func (a *Server) createOIDCClient(conn services.OIDCConnector) (*oidc.Client, er
 				"unknown problem with connector %v, most likely URL %q is not valid or not accessible, check configuration and try to re-create the connector",
 				conn.GetName(), conn.GetIssuerURL())
 		}
-		if err := a.emitter.EmitAuditEvent(ctx, &events.UserLogin{
-			Metadata: events.Metadata{
-				Type: events.UserLoginEvent,
-				Code: events.UserSSOLoginFailureCode,
-			},
-			Method: events.LoginMethodOIDC,
-			Status: events.Status{
-				Success:     false,
-				Error:       trace.Unwrap(ctx.Err()).Error(),
-				UserMessage: err.Error(),
-			},
-		}); err != nil {
-			log.WithError(err).Warn("Failed to emit OIDC login failure event.")
-		}
-		// return user-friendly error hiding the actual error in the event
-		// logs for security purposes
-		return nil, trace.ConnectionProblem(nil,
-			"failed to login with %v",
-			conn.GetDisplay())
```
Since we no longer show the users the actual login errors (but a generic login failed msg), I didn't think it was necessary to hide this error and simply return the actual error for the outer caller to emit the events (also removes double emitting)

a-palchikov commented on lib/auth/oidc.go:
```diff
@@ -242,43 +224,44 @@ func (a *Server) CreateOIDCAuthRequest(req services.OIDCAuthRequest) (*services.
-			}
+
+		if err := a.emitter.EmitAuditEvent(a.closeCtx, event); err != nil {
+			log.WithError(err).Warn("Failed to emit OIDC login failed event.")
```
nit: log these at `Debug` if these aren't critical.

kimlisa replied:
i constructed these to be consistent with others (others use warn), i think it also makes sense to use warn here b/c missing an event for auditing should stand out

a-palchikov commented on lib/web/apiserver.go:
```diff
@@ -898,14 +903,22 @@ func (h *Handler) githubLoginWeb(w http.ResponseWriter, r *http.Request, p httpr
+	logger := h.log.WithField("auth", "github")
+	logger.Debug("Console login start.")
+
+	generalErr := trace.AccessDenied(ssoLoginConsoleErr)
```
Even though this seems to reduce the error handling (if that's the intent), it will obscure the actual point of error emission and might complicate analysis as several places would be likely candidates as the error source - it's better to emit the error at each point with the corresponding `trace` error constructor.

kimlisa replied:
Revised so that each error returns its own trace err so tracing is closer to origin of actual error

russjones reviewed (approved):
Bot.

kimlisa committed:
Define trace err closer to the origin of each actual error
